cmd/copy.go

```diff
@@ -948,6 +1025,136 @@ func (cca *cookedCopyCmdArgs) processCopyJobPartOrders() (err error) {
 	return nil
 }
 
+// TODO: include this within initResourceTraverser?
+func initPipeline(ctx context.Context, location common.Location, credential common.CredentialInfo) (p pipeline.Pipeline, err error) {
+	switch location {
+	case common.ELocation.Local():
+		return nil, nil
+	case common.ELocation.Blob():
+		p, err = createBlobPipeline(ctx, credential)
+	case common.ELocation.File():
+		p, err = createFilePipeline(ctx, credential)
+	case common.ELocation.BlobFS():
+		p, err = createBlobFSPipeline(ctx, credential)
+	default:
+		err = fmt.Errorf("can't produce new pipeline for location %s", location)
+	}
+
+	return
+}
+
+func trimWildcards(path string) string {
+	if strings.Index(path, "*") == -1 {
+		return path
+	}
+
+	firstWCIndex := strings.Index(path, "*")
+	result := replacePathSeparators(path[:firstWCIndex])
+	lastSepIndex := strings.LastIndex(result, "/")
+	result = result[:lastSepIndex+1]
+
+	return result
+}
+
+func (cca *cookedCopyCmdArgs) findObject(source bool, dstIsDir bool, object storedObject) (relativePath string) {
+	var pathEncodeRules = func(path string) string {
+		loc := common.ELocation.Unknown()
+
+		if source {
+			loc = cca.fromTo.From()
+		} else {
+			loc = cca.fromTo.To()
+		}
+		pathParts := strings.Split(path, "/")
+
+		// Encode disallowed characters on windows.
+		// TODO: Inform tests of this
+		if loc == common.ELocation.Local() && runtime.GOOS == "windows" {
+			invalidChars := `<>\/:"|?*` + string(0x00)
+
+			for _, c := range strings.Split(invalidChars, "") {
+				for k, p := range pathParts {
+					pathParts[k] = strings.ReplaceAll(p, c, url.PathEscape(c))
+				}
+			}
+		} else if loc != common.ELocation.Local() {
+			for k, p := range pathParts {
+				pathParts[k] = url.PathEscape(p)
+			}
+		}
+
+		path = strings.Join(pathParts, "/")
+		return path
+	}
+
+	// source is a EXACT path to the file.
+	if object.relativePath == "" {
+		// If we're finding an object from the source, it returns "" if it's already got it.
+		// If we're finding an object on the destination and we get "", we need to hand it the object name (if it's pointing to a folder)
+		if source {
+			relativePath = ""
+		} else {
+			if dstIsDir {
+				relativePath = "/" + object.name
+			} else {
+				relativePath = ""
+			}
+		}
+
+		return pathEncodeRules(relativePath)
+	}
+
+	// If it's out here, the source is a folder (or wildcard) of some kind.
+
+	relativePath = "/" + strings.Replace(object.relativePath, common.OS_PATH_SEPARATOR, common.AZCOPY_PATH_SEPARATOR_STRING, -1)
+
+	if !source && !strings.Contains(cca.source, "*") {
```

Comment:
Can you move this (i.e. `strings.Contains(cca.source, "*") ` to a function called something like `pathPointsToContents` so that this line would become
`if !source && !pathPointsToContents(cca.source) {`

and pathPointsToContents would be something like the following?  The reason is that our /* suffix logic is sufficiently noteworthy to be wrapped in a function, so that (a) it can be clearly documented there and (b) that function can be reused if needed to become the single source of truth about this behavior and (c) by doing find-usages on that function we should, in theory, be able to find all parts of the code that have conditional logic based on this special suffix.

```
\\ pathPointsToContents differentiates between the two types of source path that we support.  Ones like this:
\\ foo\bar
\\ and ones like this:
\\ foo\bar\*
\\ The former case says to copy the whole directory (i.e. replacate bar, including the name 'bar', to the destination) whereas the latter says to only copy the contents of bar, not bar itself.
\\ NOTE: it is assumed that the path has already been validated. E.g. we don't check here for what's in front 
\\ of the *, which should be a \ or a /
func pathPointsToContents(path string) {
   return strings.HasSuffix(path, "*")      
}
```

Reply:
Will do.